### PR TITLE
Feature/fix database integration tests

### PR DESCRIPTION
**Pull request by vimanyu**
This should fix SEGFAULTS that were seen on running database integration tests. Issue was that "app_name" is usually a string with a "/" in it and mkdir calls on Mac and Windows were not creating directories recursively.

### CODE REVIEW

File: database/src/desktop/util_desktop_macos.mm [hunk 1]
```diff
@@ -17,11 +17,47 @@
...
 namespace internal {
 
+// Split a string based on specified character delimiter into constituent parts
+static std::vector<std::string> split_string(const std::string& s,
+                                             const char delimiter='/') {
+  size_t pos = 0;
+  // This index is used as the starting index to search the delimiters from.
+  size_t delimiter_search_start = 0;
+  // Skip any leading delimiters
+  while(s[delimiter_search_start] == delimiter) {
```
Comment (DellaBitta) [inline database/src/desktop/util_desktop_macos.mm:33]: Can you use std::string.find here?

File: database/src/desktop/util_desktop_macos.mm [hunk 2]
```diff
@@ -30,8 +66,17 @@
     int retval;
     retval = mkdir(path.c_str(), 0700);
     if (retval != 0 && errno != EEXIST) return "";
-    retval = mkdir((path + "/" + app_name).c_str(), 0700);
-    if (retval != 0 && errno != EEXIST) return "";
+    // App name might contain path separators. Split it to get list of subdirs
+    std::vector<std::string> app_name_parts = split_string(app_name, '/');
+    if(app_name_parts.empty()) return "";
```
Comment (DellaBitta) [inline database/src/desktop/util_desktop_macos.mm:71]: if's have spaces after if, as in "if (app_name_parts.empty()) return "";

File: database/src/desktop/util_desktop_macos.mm [hunk 3]
```diff
@@ -30,8 +66,17 @@
...
+    std::string dir_path = path;
+    // Recursively create entire tree of directories
+    for (std::vector<std::string>::const_iterator it = app_name_parts.begin();
+                                            it != app_name_parts.end(); it++) {
+          dir_path = dir_path + "/" + *it;
```
Comment (DellaBitta) [inline database/src/desktop/util_desktop_macos.mm:76]: This tabbing looks off, as if there's one too many indents.

File: database/src/desktop/util_desktop_windows.cc [hunk 4]
```diff
@@ -26,6 +26,41 @@ namespace firebase {
 namespace database {
 namespace internal {
 
+// Split a string based on specified character delimiter into constituent parts
```
Comment (DellaBitta) [inline database/src/desktop/util_desktop_windows.cc:29]: Formatting comments above apply here, too.

File: database/src/desktop/util_desktop_macos.mm [hunk 5]
```diff
@@ -17,11 +17,47 @@
...
+  }
+
+  std::vector<std::string> split_parts;
+  size_t len = s.size();
+  // Cant proceed if input string consists of just delimiters
+  if (pos >= len) {
+    return split_parts;
+  }
+
+  while( (pos = s.find(delimiter, delimiter_search_start)) != std::string::npos) {
```
Comment (DellaBitta) [inline database/src/desktop/util_desktop_macos.mm:44]: while ((

File: database/src/desktop/util_desktop_macos.mm [hunk 6]
```diff
@@ -17,11 +17,47 @@
...
+    while(s[pos] == delimiter && pos<len) {
+      pos++;
+      delimiter_search_start = pos;
+    }
+  }
+
+  // If the input string doesn't end with a delimiter we need to push the last
+  // token into our return vector
+  if (delimiter_search_start != len)
+    split_parts.push_back(s.substr(delimiter_search_start, len-delimiter_search_start));
```
Comment (DellaBitta) [inline database/src/desktop/util_desktop_macos.mm:56]: recommend putting this in brackets, like the if (pos >= len) above.

Comment (vimanyu) [reply database/src/desktop/util_desktop_macos.mm:56]: Done!

File: database/src/desktop/util_desktop_macos.mm (hunk 5 above)
Comment (vimanyu) [reply database/src/desktop/util_desktop_macos.mm]: Done!

File: database/src/desktop/util_desktop_windows.cc (hunk 4 above)
Comment (vimanyu) [reply database/src/desktop/util_desktop_windows.cc:29]: Done!

File: database/src/desktop/util_desktop_macos.mm (hunk 3 above)
Comment (vimanyu) [reply database/src/desktop/util_desktop_macos.mm]: Done!

File: database/src/desktop/util_desktop_macos.mm (hunk 2 above)
Comment (vimanyu) [reply database/src/desktop/util_desktop_macos.mm]: Done!

File: database/src/desktop/util_desktop_macos.mm (hunk 1 above)
Comment (vimanyu) [reply database/src/desktop/util_desktop_macos.mm:33]: I feel for this purpose, it will make the code real hard to follow given the edge cases.
From `find`, we can find the first position of delimiter but it could be a leading one or not. 
For the leading case, we will have to issue multiple `find`(s) repeatedly until we have found the last consecutive delimiter (Eg: ///Library/foo/bar)

Comment (DellaBitta) [reply database/src/desktop/util_desktop_macos.mm:33]: Gotcha.

**Review by DellaBitta — DISMISSED** [review]

File: database/src/desktop/util_desktop_windows.cc [hunk 7]
```diff
@@ -26,6 +26,42 @@ namespace firebase {
...
+static std::vector<std::string> split_string(const std::string& s,
```
Comment (a-maurice) [inline database/src/desktop/util_desktop_windows.cc:30]: If I'm reading this right, there is no difference between this and the above one. Should probably be moved to the shared util_desktop.cc/h

**Review by a-maurice — CHANGES_REQUESTED** [review]

File: database/src/desktop/util_desktop_windows.cc (hunk 7 above)
Comment (vimanyu) [reply database/src/desktop/util_desktop_windows.cc]: Good call. This function went from a platform dependent `mkdir` to a simple string split that should be in a common location. Moved it.

**Review by a-maurice — APPROVED** [review]
